File: examples/gimbal_manager_test/gimbal_manager_test.py
"""
Example dummy gimbal manager component, implemented in PyMAVLink.
At the current time of writing, there are some issues with implementing a dummy
gimbal manager component with the MAVLinkPassthrough module.
This example uses the libpilot MAVLink component library (https://gitlab.avhs.best/amadoruavs/pilot/pilot).

Some TODOS:
- This dummy implementation does not implement gimbal manager flags > 16.
  While this should not be a problem for pure testing purposes, it would
  probably be nice to implement those flags in some capacity.

- This gimbal manager currently DOES NOT implement deconfliction. Again, while it should work fine
  for testing, it would probably be good to implement this in the future.

Originally authored by: Vincent Wang (2020-03-30).
"""
import argparse
import time
from pilot import Component
from pilot.util import MavlinkId
from pymavlink import mavutil

class DummyGimbalManager(Component):
    def __init__(self, mavlink_url: str, mavlink_id: MavlinkId, device_id: MavlinkId):
        super().__init__(mavlink_url, mavlink_id, mavutil.mavlink.MAV_TYPE_ONBOARD_CONTROLLER)

        self.discovered = False
        self.gimbal_discovered = False
        self.starttime = time.time()
        self.device_id = device_id
        self.flags = 8
        self.gimbal_info = {"firmware_version": 0,
                            "cap_flags": 0,
                            "tilt_max": 0,
                            "tilt_min": 0,
                            "tilt_rate_max": 0,
                            "pan_max": 0,
                            "pan_min": 0,
                            "pan_rate_max": 0}

        self.register_callback(self._request_gimbal_info)
        self.register_callback(self._discover_gimbal)
        self.register_callback(self._discover_self)
        self.register_callback(self._set_attitude)
        self.register_callback(self._report_status)

        self.register_callback(self._do_gimbal_manager_attitude)

    def _activate(self, source_id):
        """
        Activate the component.
        """
        super()._activate()

    def _deactivate(self, source_id):
        """
        Deactivate the component.
        """
        super()._deactivate(source_id)

    def _request_gimbal_info(self, msg, data):
        """
        Request gimbal information as part of the gimbal device
        discovery process outlined in https://mavlink.io/en/services/gimbal_v2.html.
        Will retry at a rate of 1Hz until successful.
        """
        if not self.gimbal_discovered and ("prev_rq_time" not in data or time.time() - data["prev_rq_time"] >= 1):
            print("requesting gimbal information from", self.device_id)
            self.mavlink_conn.mav.command_long_send(self.device_id.system_id,
                                                    self.device_id.component_id,
                                                    mavutil.mavlink.MAV_CMD_REQUEST_MESSAGE,
                                                    0,
                                                    mavutil.mavlink.MAVLINK_MSG_ID_GIMBAL_DEVICE_INFORMATION,
                                                    0, 0, 0, 0, 0, 0)
            data["prev_rq_time"] = time.time()

    def _discover_gimbal(self, msg, data):
        """
        Receive gimbal capabilities and store them.
        """
        if not hasattr(msg, "id"):
            return

        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_GIMBAL_DEVICE_INFORMATION:
            info = msg.to_dict()
            print("gimbaldev", data)
            self.gimbal_info = info
            self.gimbal_discovered = True

    def _discover_self(self, msg, data):
        """
        Report gimbal capabilities as part of the gimbal manager
        discovery process outlined in https://mavlink.io/en/services/gimbal_v2.html.
        """
        if not hasattr(msg, "id"):
            return


        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_COMMAND_LONG:
            msg_data = msg.to_dict()

            # We only want to respond to a request to our specific component
            if msg_data.get("target_system") != self.mavlink_id.system_id or \
                    msg_data.get("target_component") != self.mavlink_id.component_id:
                return

            if msg.command == mavutil.mavlink.MAV_CMD_REQUEST_MESSAGE and \
                    msg_data["param1"] == mavutil.mavlink.MAVLINK_MSG_ID_GIMBAL_MANAGER_INFORMATION:
                #We don't want to be discovered until we have the gimbal device info
                if not self.gimbal_discovered:
                    self.mavlink_conn.mav.command_ack_send(msg.id, 5)
                    return
                else:
                    self.mavlink_conn.mav.command_ack_send(msg.id, 0)

                print("Sending gimbal manager information")
                self.mavlink_conn.mav.gimbal_manager_information_send(round(1000*(time.time() - self.starttime)),
                                                                      self.gimbal_info["cap_flags"],
                                                                      self.device_id.component_id,
                                                                      self.gimbal_info["tilt_max"],
                                                                      self.gimbal_info["tilt_min"],
                                                                      self.gimbal_info["tilt_rate_max"],
                                                                      self.gimbal_info["pan_max"],
                                                                      self.gimbal_info["pan_min"],
                                                                      self.gimbal_info["pan_rate_max"])

    def _set_attitude(self, msg, data):
        """
        Implement the GIMBAL_MANAGER_SET_ATTITUDE command.
        This basically just forwards the request to the gimbal device.
        Because this is a dummy device, gimbal flags > 16 are ignored and
        reset to the default (GIMBAL_DEVICE_FLAGS_PITCH_LOCK).
        """
        if not hasattr(msg, "id"):
            return

        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_GIMBAL_MANAGER_SET_ATTITUDE:
            msg_data = msg.to_dict()

            # We only want to respond to a request to our specific component
            if msg_data.get("target_system") != self.mavlink_id.system_id or \
                    msg_data.get("target_component") != self.mavlink_id.component_id:
                return

            # TODO: If we feel like it we can implement the more complicated gimbal manager flags
            # However since this is a dummy implementation anyway leaving them unimplemented and
            # falling back to default should be an acceptable approach
            self.flags = msg_data["flags"] if msg_data["flags"] <= 16 else 8
            self.mavlink_conn.gimbal_device_set_attitude_send(self.device_id.system_id,
                                                              self.device_id.component_id,
                                                              self.flags,
                                                              msg_data["q"],
                                                              msg_data["angular_velocity_x"],
                                                              msg_data["angular_velocity_y"],
                                                              msg_data["angular_velocity_z"])
            print("Set attitude! data:", msg_data)

    def _do_gimbal_manager_attitude(self, msg, data):
        """
        Implement the DO_GIMBAL_MANAGER_ATTITUDE command.
        Since this is a dummy implementation we will just print an acknowledgement.
        """
        if not hasattr(msg, "id"):
            return

        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_COMMAND_LONG:
            msg_data = msg.to_dict()

            # We only want to respond to a request to our specific component
            if msg_data.get("param1") != self.device_id.component_id:
                return

            if msg.command == mavutil.mavlink.MAV_CMD_DO_GIMBAL_MANAGER_ATTITUDE:
                print("Received DO_GIMBAL_MANAGER_ATTITUDE! data:", msg.to_dict())
                self.mavlink_conn.mav.command_ack_send(msg.id, 0)

    def _do_set_roi_location(self, msg, data):
        """
        Implement the DO_SET_ROI_LOCATION command.
        Since this is a dummy implementation we will just print an acknowledgement.
        """
        if not hasattr(msg, "id"):
            return

        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_COMMAND_LONG:
            msg_data = msg.to_dict()

            # We only want to respond to a request to our specific component
            if msg_data.get("param1") != self.device_id.component_id:
                return

            if msg.command == mavutil.mavlink.MAV_CMD_DO_SET_ROI_LOCATION:
                print("Received DO_SET_ROI_LOCATION!: data:", msg.to_dict())
                self.mavlink_conn.mav.command_ack_send(msg.id, 0)

    def _do_set_roi_wpnext_offset(self, msg, data):
        """
        Implement the DO_SET_ROI_WPNEXT_OFFSET command.
        Since this is a dummy implementation we will just print an acknowledgement.
        """
        if not hasattr(msg, "id"):
            return

        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_COMMAND_LONG:
            msg_data = msg.to_dict()

            # We only want to respond to a request to our specific component
            if msg_data.get("param1") != self.device_id.component_id:
                return

            if msg.command == mavutil.mavlink.MAV_CMD_DO_SET_ROI_WPNEXT_OFFSET:
                print("Received DO_SET_ROI_WPNEXT_OFFSET! data:", msg.to_dict())
                self.mavlink_conn.mav.command_ack_send(msg.id, 0)

    def _do_set_roi_sysid(self, msg, data):
        """
        Implement the DO_SET_ROI_SYSID command.
        Since this is a dummy implementation we will just print an acknowledgement.
        """
        if not hasattr(msg, "id"):
            return

        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_COMMAND_LONG:
            msg_data = msg.to_dict()

            # We only want to respond to a request to our specific component
            if msg_data.get("param1") != self.device_id.component_id:
                return

            if msg.command == mavutil.mavlink.MAV_CMD_DO_SET_ROI_SYSID:
                print("Received DO_SET_ROI_SYSID! data:", msg.to_dict())
                self.mavlink_conn.mav.command_ack_send(msg.id, 0)

    def _do_set_roi_none(self, msg, data):
        """
        Implement the DO_SET_ROI_NONE command.
        Since this is a dummy implementation we will just print an acknowledgement.
        """
        if not hasattr(msg, "id"):
            return

        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_COMMAND_LONG:
            msg_data = msg.to_dict()

            # We only want to respond to a request to our specific component
            if msg_data.get("param1") != self.device_id.component_id:
                return

            if msg.command == mavutil.mavlink.MAV_CMD_DO_SET_ROI_NONE:
                print("Received DO_SET_ROI_NONE! data:", msg.to_dict())
                self.mavlink_conn.mav.command_ack_send(msg.id, 0)

    def _do_gimbal_manager_track_point(self, msg, data):
        """
        Implement the DO_GIMBAL_MANAGER_TRACK_POINT command.
        Since this is a dummy implementation we will just print an acknowledgement.
        """
        if not hasattr(msg, "id"):
            return

        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_COMMAND_LONG:
            msg_data = msg.to_dict()

            # We only want to respond to a request to our specific component
            if msg_data.get("param1") != self.device_id.component_id:
                return

            if msg.command == mavutil.mavlink.MAV_CMD_DO_GIMBAL_MANAGER_TRACK_POINT:
                print("Received DO_GIMBAL_MANAGER_TRACK_POINT! data:", msg.to_dict())
                self.mavlink_conn.mav.command_ack_send(msg.id, 0)

    def _do_gimbal_manager_track_rectangle(self, msg, data):
        """
        Implement the DO_GIMBAL_MANAGER_TRACK_RECTANGLE command.
        Since this is a dummy implementation we will just print an acknowledgement.
        """
        if not hasattr(msg, "id"):
            return

        if msg.id == mavutil.mavlink.MAVLINK_MSG_ID_COMMAND_LONG:
            msg_data = msg.to_dict()

            # We only want to respond to a request to our specific component
            if msg_data.get("param1") != self.device_id.component_id:
                return

            if msg.command == mavutil.mavlink.MAV_CMD_DO_GIMBAL_MANAGER_TRACK_RECTANGLE:
                print("Received DO_GIMBAL_MANAGER_TRACK_RECTANGLE! data:", msg.to_dict())
                self.mavlink_conn.mav.command_ack_send(msg.id, 0)

    def _report_status(self, msg, data):
        """
        Report GIMBAL_MANAGER_STATUS at a rate of 5hz.
        """
        if "gm_prev_time" not in data or time.time() - data["gm_prev_time"] >= (1 / 5):
            self.mavlink_conn.mav.gimbal_manager_status_send(round(1000*(time.time() - self.starttime)),
                                                             self.flags)
            data["gm_prev_time"] = time.time()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("mavlink_url", help="URL of the MAVLink system (e.g. udp:localhost:14553).")
    parser.add_argument("gimbal_sid", help="System ID of the gimbal device")
    parser.add_argument("gimbal_cid", help="Component ID of the gimbal device")
    parser.add_argument("--manager-sid", help="System ID of the manager", default=254)
    parser.add_argument("--manager-cid", help="Component ID of the manager [25-99]", default=25)
    args = parser.parse_args()

    manager_id = MavlinkId(system_id=int(args.manager_sid), component_id=int(args.manager_cid))
    device_id = MavlinkId(system_id=int(args.gimbal_sid), component_id=int(args.gimbal_cid))

    manager = DummyGimbalManager(args.mavlink_url, manager_id, device_id)
    manager.start()

File: examples/gimbal_manager_test/gimbal_manager_test.cpp
//
// Dummy gimbal manager implementation written using MavlinkPassthrough.
// Used to test the gimbal_manager, gimbal_device, and gimbal plugins.
//
// Author: Kalyan Sriram <coder.kalyan@gmail.com>

#include <cstdint>
#include <mavsdk/mavsdk.h>
#include <mavsdk/plugins/mavlink_passthrough/mavlink_passthrough.h>
#include <iostream>
#include <algorithm>
#include <thread>
#include <sys/time.h>

using namespace mavsdk;
using namespace std::this_thread;
using namespace std::chrono;

#define ERROR_CONSOLE_TEXT "\033[31m" // Turn text on console red
#define TELEMETRY_CONSOLE_TEXT "\033[34m" // Turn text on console blue
#define NORMAL_CONSOLE_TEXT "\033[0m" // Restore normal console color

void usage(std::string bin_name)
{
    std::cout << NORMAL_CONSOLE_TEXT << "Usage : " << bin_name << " <connection_url> <id>" << std::endl
              << "Connection URL format should be :" << std::endl
              << " For TCP : tcp://[server_host][:server_port]" << std::endl
              << " For UDP : udp://[bind_host][:bind_port]" << std::endl
              << " For Serial : serial:///path/to/serial/dev:[baudrate]" << std::endl
              << "For example, to connect to the simulator use URL: udp://:14540" << std::endl
              << "Id should be from [1-6] indicating gimbal 1-6" << std::endl;
}

uint8_t parse_component_id(char* raw) {
    switch (std::stoi(raw)) {
        case 1:
            return MAV_COMP_ID_GIMBAL;
        case 2:
            return MAV_COMP_ID_GIMBAL2;
        case 3:
            return MAV_COMP_ID_GIMBAL3;
        case 4:
            return MAV_COMP_ID_GIMBAL4;
        case 5:
            return MAV_COMP_ID_GIMBAL5;
        case 6:
            return MAV_COMP_ID_GIMBAL6;
        default:
            return MAV_COMP_ID_GIMBAL;
    }
}

int main(int argc, char** argv)
{
    Mavsdk dc;
    uint8_t manager_sid = 1, manager_cid = MAV_COMP_ID_PATHPLANNER;

    Mavsdk::Configuration config(manager_cid, manager_sid);
    dc.set_configuration(config);

    std::string connection_url;
    uint8_t device_id;
    ConnectionResult connection_result;

    bool discovered_system = false;
    if (argc == 3) {
        connection_url = argv[1];
        connection_result = dc.add_any_connection(connection_url);
        device_id = parse_component_id(argv[2]);
    } else {
        usage(argv[0]);
        return 1;
    }

    std::cout << "Configured for gimbal component " << device_id
              << "(" << std::stoi(argv[2]) << ")" << std::endl;

    if (connection_result != ConnectionResult::SUCCESS) {
        std::cout << ERROR_CONSOLE_TEXT
                  << "Connection failed: " << connection_result_str(connection_result)
                  << NORMAL_CONSOLE_TEXT << std::endl;
        return 1;
    }

    System& system = dc.system();

    std::cout << "Waiting to discover system..." << std::endl;
    dc.register_on_discover([&discovered_system](uint64_t uuid) {
        std::cout << "Discovered system with UUID: " << uuid << std::endl;
        discovered_system = true;
    });

    sleep_for(seconds(2));

    if (!discovered_system) {
        std::cout << ERROR_CONSOLE_TEXT << "No system found, exiting." << NORMAL_CONSOLE_TEXT
                  << std::endl;
        return 1;
    }

    mavlink_gimbal_device_information_t gimbal_device_info;
    bool received_information = false;
    bool discovered = false;
    int manager_flags = GIMBAL_MANAGER_CAP_FLAGS_ENUM_END;

    // Here's our actual listening code
    MavlinkPassthrough pass(system);

    struct timeval tv_start;
    struct timeval tv_status_last;
    struct timeval tv_hb_last;
    struct timeval tv;
    unsigned long time_in_mils;
    unsigned long time_status_last;
    unsigned long time_hb_last;
    unsigned long time_start;

    // Discover our gimbal device
    pass.subscribe_message_async(MAVLINK_MSG_ID_GIMBAL_DEVICE_INFORMATION,
        [&received_information, &gimbal_device_info, &pass](const mavlink_message_t message) {
        mavlink_msg_gimbal_device_information_decode(&message, &gimbal_device_info);
        received_information = true;
    });

    // Listen for commands
    pass.subscribe_message_async(MAVLINK_MSG_ID_COMMAND_LONG,
        [&received_information, &discovered, &device_id, &manager_sid, &manager_cid, &pass, &gimbal_device_info, &time_in_mils, &time_start](const mavlink_message_t message) {

        mavlink_command_long_t cmd;
        mavlink_msg_command_long_decode(&message, &cmd);

        uint16_t command_types[] = { MAV_CMD_DO_GIMBAL_MANAGER_ATTITUDE,
                                     MAV_CMD_DO_SET_ROI_LOCATION,
                                     MAV_CMD_DO_SET_ROI_WPNEXT_OFFSET,
                                     MAV_CMD_DO_SET_ROI_SYSID,
                                     MAV_CMD_DO_SET_ROI_NONE,
                                     MAV_CMD_DO_GIMBAL_MANAGER_TRACK_POINT,
                                     MAV_CMD_DO_GIMBAL_MANAGER_TRACK_RECTANGLE };
        const char* command_names[] = { "MAV_CMD_DO_GIMBAL_MANAGER_ATTITUDE",
                                        "MAV_CMD_DO_SET_ROI_LOCATION",
                                        "MAV_CMD_DO_SET_ROI_WPNEXT_OFFSET",
                                        "MAV_CMD_DO_SET_ROI_SYSID",
                                        "MAV_CMD_DO_SET_ROI_NONE",
                                        "MAV_CMD_DO_GIMBAL_MANAGER_TRACK_POINT",
                                        "MAV_CMD_DO_GIMBAL_MANAGER_TRACK_RECTANGLE" };
        short unsigned int* command_index = std::find(command_types, command_types+7, cmd.command);

        if (cmd.command == MAV_CMD_REQUEST_MESSAGE) {
            std::cout << "Received request for gimbal manager info!" << std::endl;
            if (cmd.param1 == MAVLINK_MSG_ID_GIMBAL_MANAGER_INFORMATION) {
                // Respond to requests with gimbal manager info

                // If we're already discovered, don't respond
                if (discovered) return;

                if (!received_information) {
                    std::cout << "Gimbal device info not yet received, sending in progress message" << std::endl;
                    mavlink_message_t ack;
                    mavlink_msg_command_ack_pack(manager_sid, manager_cid, &ack,
                        MAV_CMD_REQUEST_MESSAGE, MAV_RESULT_IN_PROGRESS, 0, 0, 0, 0);

                    pass.send_message(ack);
                } else {
                    mavlink_message_t ack;
                    mavlink_msg_command_ack_pack(manager_sid, manager_cid, &ack,
                        MAV_CMD_REQUEST_MESSAGE, MAV_RESULT_ACCEPTED, 0, 0, 0, 0);
                    pass.send_message(ack);

                    mavlink_message_t info;
                    mavlink_msg_gimbal_manager_information_pack(manager_sid, manager_cid,
                            &info,
                            time_in_mils - time_start,
                            manager_sid, device_id,
                            gimbal_device_info.tilt_max, gimbal_device_info.tilt_min,
                            gimbal_device_info.tilt_rate_max,
                            gimbal_device_info.pan_max, gimbal_device_info.pan_min,
                            gimbal_device_info.pan_rate_max);
                    pass.send_message(info);
                    std::cout << "Sent gimbal info!" << std::endl;
                    discovered = true;
                }
            }
        } else if (command_index != command_types+7) {
            // Respond to various gimbal manager commands
            // Since this is a dummy we just print an acknowledgement

            // We only want to respond to requests for our gimbal device; if it's
            // not ours, ignore
            // Until SET_ATTITUDE gets a proper device id field we'll have to settle
            // for testing against component and system id
            if (cmd.target_system != manager_sid || cmd.target_component != manager_cid) return;

            std::cout << "Received " <<  command_names[command_index - command_types] << ", data:"
                      << cmd.param1 << ", "
                      << cmd.param2 << ", "
                      << cmd.param3 << ", "
                      << cmd.param4 << ", "
                      << cmd.param5 << ", "
                      << cmd.param6 << ", "
                      << cmd.param7 << std::endl;

            // Reply with COMMAND_ACK
            mavlink_message_t ack;
            mavlink_msg_command_ack_pack(manager_sid, manager_cid,
                    &ack,
                    cmd.command,
                    MAV_RESULT_ACCEPTED,
                    0, 0, message.sysid, message.compid);
            pass.send_message(ack);
        }
    });

    pass.subscribe_message_async(MAVLINK_MSG_ID_GIMBAL_MANAGER_SET_ATTITUDE,
        [&device_id, &manager_sid, &manager_cid, &pass](const mavlink_message_t message) {
        mavlink_gimbal_manager_set_attitude_t set_attitude_manager;
        mavlink_msg_gimbal_manager_set_attitude_decode(&message, &set_attitude_manager);

        // We only want to respond to requests for our gimbal device; if it's
        // not ours, ignore

        if (set_attitude_manager.flags >= 16) {
            // Override manager flags to 8 (default) if flags is greater than 16
            // (normally, we would implement 16+ in the gimbal manager control code)
            // (however, since this is a dummy manager, it doesn't make sense to)
            set_attitude_manager.flags = 8;
        }

        mavlink_message_t set_attitude_device;
        mavlink_msg_gimbal_device_set_attitude_pack(manager_sid, manager_cid,
                &set_attitude_device,
                manager_cid, device_id,
                set_attitude_manager.flags,
                set_attitude_manager.q,
                set_attitude_manager.angular_velocity_x,
                set_attitude_manager.angular_velocity_y,
                set_attitude_manager.angular_velocity_z);
        pass.send_message(set_attitude_device);

        std::cout << "Set attitude! ("
                  << set_attitude_manager.angular_velocity_x << ", "
                  << set_attitude_manager.angular_velocity_y << ", "
                  << set_attitude_manager.angular_velocity_z << ")" << std::endl;
    });

    gettimeofday(&tv_start, NULL);
    gettimeofday(&tv_status_last, NULL);
    gettimeofday(&tv_hb_last, NULL);
    time_start = 1000 * tv_start.tv_sec + (int) (tv_start.tv_usec * 0.001);

    // publish messages regularly
    while (true) {
        gettimeofday(&tv, NULL);
        time_in_mils = 1000 * tv.tv_sec + (int) (tv.tv_usec * 0.001);
        time_status_last = 1000 * tv_status_last.tv_sec + (int) (tv_status_last.tv_usec * 0.001);
        time_hb_last = 1000 * tv_hb_last.tv_sec + (int) (tv_hb_last.tv_usec * 0.001);

        // Send GIMBAL_MANAGER_STATUS messages at a rate of 5hz (every 200 ms)
        if (time_in_mils - time_status_last > 200) {
            mavlink_message_t status_message;
            mavlink_msg_gimbal_manager_status_pack(manager_sid, manager_cid,
                    &status_message, time_in_mils - time_start, manager_flags);
            pass.send_message(status_message);
            gettimeofday(&tv_status_last, NULL);
        }

        // Send HEARTBEAT messages at a rate of 1hz (every 1000ms)
        if (time_in_mils - time_hb_last > 1000) {
            mavlink_message_t hb_message;
            mavlink_msg_heartbeat_pack(manager_sid, manager_cid,
                    &hb_message, MAV_TYPE_ONBOARD_CONTROLLER,
                    MAV_AUTOPILOT_INVALID,
                    0, 0, MAV_STATE_ACTIVE);
            pass.send_message(hb_message);

            // Request gimbal info at a rate of 1hz if we don't have it yet
            if (!received_information) {
                std::cout << "Requesting gimbal device info" << std::endl;
                mavlink_message_t request_message;
                mavlink_msg_command_long_pack(manager_sid, manager_cid,
                        &request_message,
                        manager_sid, device_id,
                        MAV_CMD_REQUEST_MESSAGE,
                        0,
                        MAVLINK_MSG_ID_GIMBAL_DEVICE_INFORMATION,
                        0, 0, 0, 0, 0, 0);
            }
            gettimeofday(&tv_hb_last, NULL);
        }

        sleep_for(milliseconds(50));
    }

    return 0;
}
